mayastor/src/replicas/rebuild/rebuild_impl.rs

```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
+        }
+    }
+
+    fn pending_not(&self, state: RebuildState) -> bool {
+        self.pending.is_some() && self.pending != Some(state)
+    }
+
+    fn pending_is(&self, state: RebuildState) -> bool {
+        self.pending == Some(state)
+    }
+
+    fn reconcile(&mut self) -> RebuildState {
+        if let Some(pending) = self.pending {
+            self.current = pending;
+            self.pending = None;
+        }
+
+        self.current
+    }
+}
+
+impl RebuildJob {
+    /// Client operations are now allowed to skip over previous operations
+    fn exec_client_op(
+        &mut self,
+        op: RebuildOperation,
+    ) -> Result<(), RebuildError> {
+        self.exec_op(op, false)
+    }
+    fn exec_internal_op(
+        &mut self,
+        op: RebuildOperation,
+    ) -> Result<(), RebuildError> {
+        self.exec_op(op, true)
+    }
+
+    /// Single state machine where all operations are handled
+    fn exec_op(
+        &mut self,
+        op: RebuildOperation,
+        override_pending: bool,
+    ) -> Result<(), RebuildError> {
+        type S = RebuildState;
+        let e = RebuildError::OpError {
+            operation: op.to_string(),
+            state: self.states.to_string(),
+        };
+
+        trace!(
+            "Executing operation {} with override {}",
+            op,
+            override_pending
+        );
+
+        match op {
+            RebuildOperation::Start => {
+                match self.state() {
+                    // start only allowed when... starting
+                    S::Stopped | S::Paused | S::Failed | S::Completed => Err(e),
+                    // for idempotence sake
+                    S::Running => Ok(()),
+                    S::Init => {
+                        self.states.set_pending(S::Running, false)?;
+                        self.schedule();
+                        Ok(())
+                    }
+                }
+            }
+            RebuildOperation::Stop => {
+                match self.state() {
+                    // We're already stopping anyway, so all is well
+                    S::Failed | S::Completed => Err(e),
+                    // for idempotence sake
+                    S::Stopped => Ok(()),
+                    S::Running => {
+                        self.states
+                            .set_pending(S::Stopped, override_pending)?;
+                        Ok(())
+                    }
+                    S::Init | S::Paused => {
+                        self.states
+                            .set_pending(S::Stopped, override_pending)?;
+
+                        // The rebuild is not running so we need to notify here
+                        self.notify();
+                        Ok(())
+                    }
+                }
+            }
+            RebuildOperation::Pause => match self.state() {
+                S::Stopped | S::Failed | S::Completed => Err(e),
+                S::Init | S::Running | S::Paused => {
```

Comment:
Should the paused state be split out on its own so it resembles the other cases i.e. "for idempotence sake" on line 584?

Reply:
Done